### DISCUSSION TITLE
How to handle handwritten German text vs printed text?

### DISCUSSION THREAD

**Discussion by DefcoGit** [Keyvanhardani/german-ocr#11] [Q&A] [answered]
Hello,

I am building a document processing pipeline and need to handle both:
- Printed German text (invoices, contracts)
- Handwritten notes and signatures

**Issues I am facing:**
1. The OCR accuracy drops significantly for handwritten content
2. Mixed documents (printed + handwritten) are problematic
3. Old German fonts (Fraktur) are not recognized well

**My current approach:**
```python
result = ocr_engine.process(image)
# No distinction between handwritten/printed
```

Is there a recommended way to:
1. Detect handwritten vs printed regions?
2. Apply different processing strategies?
3. Handle Fraktur/old German fonts?

Thanks in advance!

**Comment by Keyvanhardani** [comment, marked as answer]
Excellent question! Handling mixed content is one of the trickier aspects of German OCR. Here is my approach:

## 1. Detecting Handwritten vs Printed Regions

Use a two-stage detection pipeline:

```python
import cv2
import numpy as np

def classify_text_region(image_region):
    # Analyze stroke variation - handwriting has more variance
    gray = cv2.cvtColor(image_region, cv2.COLOR_BGR2GRAY)
    
    # Calculate stroke width variation
    edges = cv2.Canny(gray, 50, 150)
    stroke_var = np.std(edges[edges > 0])
    
    # Handwriting typically has higher variation
    return "handwritten" if stroke_var > 45 else "printed"
```

## 2. Different Processing Strategies

```python
def process_mixed_document(image):
    regions = detect_text_regions(image)
    results = []
    
    for region in regions:
        region_type = classify_text_region(region)
        
        if region_type == "printed":
            # Use standard OCR with high confidence threshold
            result = ocr_engine.process(region, confidence=0.85)
        else:
            # Handwriting: use specialized model + lower threshold
            result = handwriting_model.process(region, confidence=0.6)
            result.is_handwritten = True
        
        results.append(result)
    
    return merge_results(results)
```

## 3. Fraktur/Old German Fonts

For Fraktur, I recommend:

```python
# Option A: Use specialized Fraktur model
from german_ocr.models import FrakturOCR

fraktur_engine = FrakturOCR()
result = fraktur_engine.process(old_document)

# Option B: Pre-process to normalize
def normalize_fraktur(image):
    # Increase contrast for old documents
    clahe = cv2.createCLAHE(clipLimit=2.0, tileGridSize=(8,8))
    enhanced = clahe.apply(cv2.cvtColor(image, cv2.COLOR_BGR2GRAY))
    
    # Denoise aged paper
    denoised = cv2.fastNlMeansDenoising(enhanced, h=10)
    return denoised
```

## Pro Tips

- **Signatures**: Dont try to OCR them - just detect and mark as "signature region"
- **Confidence scoring**: Always return confidence scores for downstream filtering
- **Training data**: If accuracy is critical, fine-tune on your specific document types

Let me know if you need more specific guidance for your use case!
